File: rx-location/src/main/java/com/nicolasmilliard/rxlocation/LocationResultObservable.java
package com.nicolasmilliard.rxlocation;

import android.annotation.SuppressLint;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.tasks.Task;
import com.nicolasmilliard.rxtask.ObservableTask;
import com.nicolasmilliard.rxtask.ObservableTaskCallback;

final class LocationResultObservable extends ObservableTask<LocationResult> {
    private final FusedLocationProviderClient client;
    private final LocationRequest request;

    public LocationResultObservable(FusedLocationProviderClient client, LocationRequest request) {
        this.client = client;
        this.request = request;
    }

    @SuppressLint("MissingPermission")
    @Override
    protected Task<Void> run(ObservableTaskCallback<LocationResult> callback) {
        ResultCallback resultCallback = new ResultCallback(callback);
        callback.setDisposeListener(() -> client.removeLocationUpdates(resultCallback));
        return client.requestLocationUpdates(this.request, resultCallback, null)
                .addOnCompleteListener(callback);
    }

    static final class ResultCallback extends LocationCallback {

        final ObservableTaskCallback<LocationResult> callback;

        public ResultCallback(ObservableTaskCallback<LocationResult> callback) {
            this.callback = callback;
        }

        @Override
        public void onLocationResult(LocationResult result) {
            super.onLocationResult(result);
            callback.onNext(result);
        }
    }
}

File: rx-location/src/main/java/com/nicolasmilliard/rxlocation/LocationResultObservable.kt
package com.nicolasmilliard.rxlocation

import android.annotation.SuppressLint
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.exceptions.CompositeException
import io.reactivex.exceptions.Exceptions
import io.reactivex.plugins.RxJavaPlugins


class LocationResultObservable(private val client: FusedLocationProviderClient, private val request: LocationRequest) : Observable<LocationResult>() {

    @SuppressLint("MissingPermission")
    override fun subscribeActual(observer: Observer<in LocationResult>) {
        val callback = ResultCallback(client, observer)
        observer.onSubscribe(callback)
        client.requestLocationUpdates(request, callback, null).addOnCompleteListener(callback)
    }

    internal class ResultCallback(private val client: FusedLocationProviderClient, private val observer: Observer<in LocationResult>) : LocationCallback(), Disposable, OnCompleteListener<Void> {
        override fun onLocationResult(result: LocationResult) {
            super.onLocationResult(result)
            if (disposed) return
            observer.onNext(result)
        }

        private var disposed = false

        override fun onComplete(task: Task<Void>) {
            if (disposed) return
            if (!task.isSuccessful) {
                try {
                    observer.onError(task.exception!!)
                } catch (inner: Throwable) {
                    Exceptions.throwIfFatal(inner)
                    RxJavaPlugins.onError(CompositeException(task.exception, inner))
                }
            }
        }

        override fun isDisposed(): Boolean = disposed

        override fun dispose() {
            disposed = true
            client.removeLocationUpdates(this).addOnCompleteListener({
                // Not sure what to do here
            })
        }

    }
}